# Computer Lab Exercises week 8

* These lab exercises are designed to further develop you skills with working with pandas dataframes
* These exercises are **not** part of the assessment of the Module.

* I will post the solutions roughly 1 week after this file has been uploaded to the DLE.
* Please first watch and work through the linkedin learning video   **Pandas Essential Training**
* See https://sites.google.com/view/pythonforstemstudents/home for a collection
of links to information on python.

see this link about pandas https://www.w3schools.com/python/pandas/


## Question 1
The dataframe below comes from a simulated medical study of some students at University. Use
pandas to do the following tasks.

*  Delete the **Student_ID**  column
*  Change the age of Sahil from 41 to 21
*  Compute the mean weight. 
*  Replace the unknown weights wih the mean weight.

The dataframe is in the cell below

In [1]:
import pandas as pd
import numpy as np

data=[["Karan",23,80,2034],["Rohit",22,np.nan,678],["Sahil",41,95,673],["Aryan",22,np.nan, 1786]]
df=pd.DataFrame(data,columns=["Name","Age", "Weight_kg", "Student_ID"])

print(df)

    Name  Age  Weight_kg  Student_ID
0  Karan   23       80.0        2034
1  Rohit   22        NaN         678
2  Sahil   41       95.0         673
3  Aryan   22        NaN        1786


The output should like:
<pre>
   Name  Age  Weight_kg  Student_ID
0  Karan   23       80.0        2034
1  Rohit   22        NaN         678
2  Sahil   41       95.0         673
3  Aryan   22        NaN        1786
Delete Student_ID row
    Name  Age  Weight_kg
0  Karan   23       80.0
1  Rohit   22        NaN
2  Sahil   41       95.0
3  Aryan   22        NaN
Update age of Sahil 
    Name  Age  Weight_kg
0  Karan   23       80.0
1  Rohit   22        NaN
2  Sahil   21       95.0
3  Aryan   22        NaN
Mean weight =  87.5 kg
    Name  Age  Weight_kg
0  Karan   23       80.0
1  Rohit   22       87.5
2  Sahil   21       95.0
3  Aryan   22       87.5

</pre>

In [2]:
##  put your solution here




##  Question 2


The Devon and Cornwall police provide data about stop
and search between 2018 and 2019 at:
https://plymouth.thedata.place/dataset/plymouth-stop-search-data

* Use pandas to delete the following columns: 'Placestoppedstreet' , 'Monthyear' , 'GRIDREFERENCE', 'GRIDREF1' and 'GRIDREF2'.

* Use pandas to compute the maximum and minimum 'EstimatedAge'.

* Delete any rows with ages which do not look correct.

* Sometimes PLYMOUTH, Plmyouth, and Plymouth is used in the data set. Use the **replace** method to  change PLYMOUTH, Plmyouth, to Plymouth.

* Compute the mean age for each area.


The final output should look like:
<pre>
                  EstimatedAge
Placestoppedtown              
Devenport            41.000000
Honiton              35.000000
Newton Abbot         34.000000
Plymouth             31.905882
Plympton             55.000000
St. Ives             18.000000
Tavistock            34.000000
Truro                41.000000
</pre>


In [3]:
import pandas as pd

# The file is available on the DLE
df = pd.read_csv("resources_57e2507e-7342-4cb1-b284-014caaa463a2_extract-for-dataplay-dataplay-master-csv.csv")
print(df.head())

print(df.columns)

  Placestoppedstreet Placestoppedtown  EstimatedAge Monthyear   GRIDREFERENCE  \
0                NaN              NaN            17    Apr-19  244974; 060033   
1                NaN              NaN            16    May-19  248200; 060630   
2                NaN              NaN            14    Sep-19  246933; 056200   
3                NaN              NaN            35    Aug-19  248182; 054717   
4                NaN              NaN            25    May-19  246332; 059320   

   GRIDREF1  GRIDREF2  
0  244974.0   60033.0  
1  248200.0   60630.0  
2  246933.0   56200.0  
3  248182.0   54717.0  
4  246332.0   59320.0  
Index(['Placestoppedstreet', 'Placestoppedtown', 'EstimatedAge', 'Monthyear',
       'GRIDREFERENCE', 'GRIDREF1', 'GRIDREF2'],
      dtype='object')


In [4]:
## Put your solution here



##  Question 3: python programming


This is an example of python programming. The file 
life-expectancy-females-2010-2014.csv contains some
information about life expectancy of women in Plymouth in
different areas of Plymouth.
The data came from https://plymouth.thedata.place/dataset/life-expectancy-data-plymouth.

Many machine learning algorithms require that the input data 
(in this case life time data) is
scaled to be range 0 to 1,  or the range -1 to 1.

If the input data is $x_i$ then the **normalization scaling** is
$$
\overline{x}_i = \frac{x_i - min} {max - min}
$$
where $min$ is the minimum value and $max$ is the
maximum value.

Another type of scaling is **standardization.**
$$
\overline{y}_i = \frac{x_i - \mu} {\sigma}
$$
where
$\mu$ is the mean of the data and
$\sigma$ is the standard deviation of the data.

* Compute the normalization scaling for the life expectancy data.
* Compute the standardization scaling for the life expectancy data.

In [5]:
import pandas as pd
import numpy as np

# the file is avilable on the DLE
df = pd.read_csv("life-expectancy-females-2010-2014.csv")
print(df.head())
print(df.columns)

life = df['Life expectancy at birth for females, 2010-2014 (years)'].tolist()
for x in life:
   print(x)


   Ward Code Plymouth Wards (2015)  \
0  E05002096              Budshead   
1  E05002094               Compton   
2  E05002080             Devonport   
3  E05002081                 Drake   
4  E05002085     Efford and Lipson   

   Life expectancy at birth for females, 2010-2014 (years)  \
0                                               78.8         
1                                               79.0         
2                                               79.3         
3                                               79.4         
4                                               81.0         

   LCL - Lower confidence limit  UCL - Upper confidence limit  
0                          77.5                          80.1  
1                          77.3                          80.7  
2                          77.9                          80.6  
3                          76.7                          82.1  
4                          79.7                          82.3  
Index(['Ward C

In [6]:
# Put your solution here 



You can read more about scaling of input data in chapter 5 of the 
100 page machine learning book 
(http://themlbook.com/)
You can download chapters in pdf from this site.

Use pandas to plot a histogram of the 'Life expectancy at birth for females, 2010-2014 (years)'


 <img src="https://github.com/cmcneile/COMP5000-2023-lab/blob/main/lifehist.png?raw=true">

In [7]:
# Put your solution here

##  Question 4 : using pandas

There is data about air quality around hospitals and
surgeries available at:
https://plymouth.thedata.place/dataset/air-quality-data.
The key column is "PM2\_5" which is the density of
small particles, which can enter the blood stream via 
the lungs. The World Health organization warns that
levels of ``PM2\_5'' above 10 can cause health problems
(see https://www.blf.org.uk/take-action/campaign/nhs-toxic-air-report.)




In [8]:
import pandas as pd

df = pd.read_csv("resources_cd162ad1-d7d5-42a9-b1ab-0edbcd697f1e_air-quality-by-pm2.5-score-blf.org.uk.csv")
print(df.head())

print(df.columns)

                       Name  \
0          West Hoe Surgery   
1   Adelaide Street Surgery   
2   Devonport Health Centre   
3            Armada Surgery   
4  North Road West Med.ctr.   

                                             Address Post Code  \
0    West Hoe Surgery, 2 Cliff Road, Plymouth, Devon   PL1 3BP   
1    20 Adelaide Street, Stonehouse, Plymouth, Devon   PL1 3JF   
2          Damerel Close, Devonport, Plymouth, Devon   PL1 4JZ   
3  Armada Surg, 28 Oxford Pl, Western Approach, P...   PL1 5AJ   
4  North Road West Med.ctr., 167 North Road West,...   PL1 5BZ   

                          Town         Type  PM2_5  Exceed_10   Latitude  \
0  St Peter and the Waterfront  GP Practice   8.87      False  50.364834   
1  St Peter and the Waterfront  GP Practice  10.32       True  50.370562   
2                    Devonport  GP Practice  10.71       True  50.370004   
3  St Peter and the Waterfront  GP Practice  10.80       True  50.374693   
4  St Peter and the Waterfront  GP P

Use pandas to answer the following questions:

* **(a)** What is the mean value of "PM2\_5" for all the data in the file?

* **(b)** Use the groupby pandas method to find the total number of   "GP Practice" and "Hospital".

* **(c)** Print the "Latitude" and "Longitude" of the locations with "PM2\_5" greater than 10.

<pre>
Answer Q4a
Mean PM2_5  =  9.227671232876713
Answer Q4b
Type
GP Practice    61
Hospital       12
Name: Name, dtype: int64
--------------------------------------------------
Answer Q4c
     Latitude  Longitude
1   50.370562  -4.155762
2   50.370004  -4.167845
3   50.374693  -4.144895
4   50.376062  -4.146265
20  50.383921  -4.177328
21  50.387979  -4.160899
22  50.393672  -4.147820
31  50.344309  -4.760374
38  50.379001  -4.166779
39  50.389271  -4.129838
46  50.368649  -4.122853
47  50.382894  -4.139146
48  50.382512  -4.131575
49  50.378837  -4.116613
50  50.372660  -4.128292
51  50.373198  -4.127909
52  50.404946  -4.188552
54  50.410582  -4.139943
55  50.408630  -4.134098
56  50.411256  -4.159425
57  50.417443  -4.150289

</pre>


In [9]:
##  Put your solution  here



## Question 5 

The code below creates an interactive plot using plotly https://dash.plotly.com/minimal-app
The application loads in a spreadsheet https://raw.githubusercontent.com/plotly/datasets/master/gapminder_unfiltered.csv

To run the application you need to install dash. Type the command below in a cell

pip install dash



In [10]:
from dash import Dash, html, dcc, callback, Output, Input
import plotly.express as px
import pandas as pd

df = pd.read_csv('https://raw.githubusercontent.com/plotly/datasets/master/gapminder_unfiltered.csv')

app = Dash()

app.layout = [
    html.H1(children='Title of Dash App', style={'textAlign':'center'}),
    dcc.Dropdown(df.country.unique(), 'Canada', id='dropdown-selection'),
    dcc.Graph(id='graph-content')
]

@callback(
    Output('graph-content', 'figure'),
    Input('dropdown-selection', 'value')
)
def update_graph(value):
    dff = df[df.country==value]
    return px.line(dff, x='year', y='pop')

if __name__ == '__main__':
    app.run(debug=True)


The question is to modify the above example to load in the spreadsheet  
https://calmcode.io/static/data/bigmac.csv
This has data for the price of bigmac as a function of date for different counteries.

Use dash to plot 

dollar_price versus date

with name in the drop down box.

The solution should look like

![dashEG](https://github.com/cmcneile/COMP5000-2024-lab/blob/main/dash.png?raw=true)

